lib/dns/server.js

```diff
@@ -449,6 +454,10 @@ class RootServer extends DNSServer {
     rd.typeBitmap = TYPE_MAP;
     return rr;
   }
+
+  toKEY() {
+    return hsig.makeKey(this.key);
```

Comment:
So this works with `dig`:

```
    const rr = hsig.makeKey(this.key);
    rr.data.algorithm = 4;
    return rr;
```

I suspect the `PRIVATEDNS` experimental algorithm code is not standardized, so `dig` parses it as a compression byte and fails.

Reply:
So, it appears even with this we see:

`;; Warning: Message parser reports malformed message packet.`

Also, `PRIVATEDNS` is defined in RFC 4034: https://tools.ietf.org/html/rfc4034 so it should be considered standard. I believe a combination of the warning above and a invalid case in parsing leads to the error described above.

Reply:
OK, so `dig` requires that the class in the answer section match the question section. So

`rr.class = classes.IN;`

solves the warning. The issue with parsing still exists.

Reply:
So, these are the offending lines:

	if (algorithm == DNS_KEYALG_PRIVATEDNS) {
		dns_name_t name;
		dns_decompress_setmethods(dctx, DNS_COMPRESS_NONE);
		dns_name_init(&name, NULL);
		RETERR(dns_name_fromwire(&name, source, dctx, options, target));
	}

Is there a spec for how to handle `PRIVATEDNS`? Not sure what's going on here...

But if we are planning to use a different key other than the brontide key anyway, why not just use a key of the type `ECDSAP256SHA256` like `KSK` and `ZSK`... that would avoid the experimental `PRIVATEDNS` case altogether which seems to be under/unspecified.

Reply:
Alright, so RFC 4034 says:

> Algorithm number 253 is reserved for private use and will never be assigned to
> a specific algorithm.  The public key area in the DNSKEY RR and the signature
> area in the RRSIG RR begin with a wire encoded domain name, which MUST NOT be
> compressed.  The domain name indicates the private algorithm to use, and the
> remainder of the public key area is determined by that algorithm.  Entities
> should only use domain names they control to designate their private
> algorithms.


https://tools.ietf.org/html/rfc4034#appendix-A.1.1

So for `PRIVATEDNS` algorithm, `dig` actually tries to parse the pubkey part of the `KEY` record as a algo name string first, followed by the actual pubkey. However, the code that parses this is interleaved with the compression decoding code (even though RFC says name will never be compressed), causing a `bad label` error if any byte falls in the range 128 - 192.

As a very simple test case to verify this, if you concat the pubkey with a null byte, you can see that `dig` is able to parse the result without issues.